### PR TITLE
db_stress: generate the key based on Zipfian distribution (hot key)

### PR DESCRIPTION
**Pull request by zhichao-cao**
In the current db_stress, all the keys are generated randomly and follows the uniform distribution. In order to test some corner cases that some key are always updated or read, we need to generate the key based on other distributions. In this PR, the key is generated based on Zipfian distribution and the skewness can be controlled by setting hot_key_alpha (0.8 to 1.5 is suggested). The larger hot_key_alpha is, the more skewed will be. Not that, usually, if hot_key_alpha is larger than 2, there might be only 1 or 2 keys that are generated. If hot_key_alpha is 0, it generate the key follows uniform distribution (random key)

Testing plan: pass the db_stress and printed the keys to make sure it follows the distribution.

### CODE REVIEW

File: db_stress_tool/db_stress_common.cc [hunk 1]
```diff
@@ -9,15 +9,57 @@
...
     rocksdb::kSnappyCompression;
 enum rocksdb::ChecksumType FLAGS_checksum_type_e = rocksdb::kCRC32c;
 enum RepFactory FLAGS_rep_factory = kSkipList;
+std::vector<double> sum_probs(100001);
+int64_t zipf_sum_size = 100000;
 
 namespace rocksdb {
+
+// Zipfian distribution is generated based on a pre-calculated array.
+// It should be used before start the stress test
```
Comment (siying) [inline db_stress_tool/db_stress_common.cc]: Can you document the Math behind this algorithm in the comment?

File: db_stress_tool/db_stress_common.cc [hunk 2]
```diff
@@ -9,15 +9,73 @@
...
+
+// Generate one key that follows the Zipfian distribution. The skewness
+// is decided by the parameter alpha. Input is the rand_seed [0,1] and
+// the max of the key to be generated. If we directly return tmp_zipf_seed,
+// the closer to 0, the higher probability will be. To randomly distribute
+// the hot keys in [0, max_key], we use Random64 to shuffle it.
+int64_t GetOneHotKeyID(double rand_seed, int64_t max_key) {
+  int64_t low = 1, mid, high = zipf_sum_size, zipf = 0;
+  while (low <= high) {
+    mid = floor((low+high)/2);
```
Comment (siying) [inline db_stress_tool/db_stress_common.cc]: Did we use namespace std? Why we can use floor without std::floor? Or are we using another floor()?

Btw, run "make format" to format the line.

File: db_stress_tool/db_stress_common.cc [hunk 3]
```diff
@@ -70,7 +128,16 @@ int64_t GenerateOneKey(ThreadState* thread, uint64_t iteration) {
...
       completed_ratio * (FLAGS_max_key - FLAGS_active_width));
-  return base_key + thread->rand.Next() % FLAGS_active_width;
+  int64_t rand_seed = base_key + thread->rand.Next() % FLAGS_active_width;
+  int64_t cur_key = rand_seed;
+  if (FLAGS_hot_key_alpha != 0) {
+    // If set the Zipfian distribution Alpha to non 0, use Zipfian
+    double float_rand =
+        (static_cast<double>(thread->rand.Next() % FLAGS_max_key)) /
+        FLAGS_max_key;
+    cur_key = GetOneHotKeyID(float_rand, FLAGS_max_key);
```
Comment (siying) [inline db_stress_tool/db_stress_common.cc]: Shouldn't we still do the base_key and FLAGS_active_width thing for hot keys?

File: db_stress_tool/db_stress_common.cc [hunk 4]
```diff
@@ -84,9 +151,17 @@ std::vector<int64_t> GenerateNKeys(ThreadState* thread, int num_keys,
...
+    // Generate the key follows zipfian distribution
+    if (FLAGS_hot_key_alpha != 0) {
+      double float_rand =
+          (static_cast<double>(thread->rand.Next() % FLAGS_max_key)) /
+          FLAGS_max_key;
+      next_key = GetOneHotKeyID(float_rand, FLAGS_max_key);
+    } else {
+      // This may result in some duplicate keys
+      next_key = next_key + thread->rand.Next() %
+                                (FLAGS_active_width - (next_key - base_key));
```
Comment (siying) [inline db_stress_tool/db_stress_common.cc]: It used to be that, the keys are sorted. But randomized keys aren't ordered. Is it a problem?

File: db_stress_tool/db_stress_common.cc [hunk 5]
```diff
@@ -84,9 +151,17 @@ std::vector<int64_t> GenerateNKeys(ThreadState* thread, int num_keys,
...
   keys.push_back(next_key);
   for (int i = 1; i < num_keys; ++i) {
-    // This may result in some duplicate keys
-    next_key = next_key + thread->rand.Next() %
-                              (FLAGS_active_width - (next_key - base_key));
+    // Generate the key follows zipfian distribution
+    if (FLAGS_hot_key_alpha != 0) {
+      double float_rand =
+          (static_cast<double>(thread->rand.Next() % FLAGS_max_key)) /
+          FLAGS_max_key;
```
Comment (siying) [inline db_stress_tool/db_stress_common.cc]: Same as above. Is it possible to follow the active_width?

File: db_stress_tool/db_stress_common.cc (hunk 3 above)
Comment (zhichao-cao) [reply db_stress_tool/db_stress_common.cc]: @siying I think we cannot use these. base_key is like a starting point of a key range, from 0 move to (max_key - active_width) based on the completion ratio. Keys are randomly generated in the key range. So overall, the probability of each key being accessed are the same. However, if we use the active_width and do Zipfian in each range, the probability of each key being access are accumulated as the "keyrange window" slides. Statistically, each key is accessed with the same probability. If we do batch by batch (one batch of accesses only happen in a active_width key range), the behavior will be strange.

File: db_stress_tool/db_stress_common.cc (hunk 4 above)
Comment (zhichao-cao) [reply db_stress_tool/db_stress_common.cc]: @siying The comment is the original one. From my understanding, the next_key might be duplicate in the array since when the range shrinks (FLAGS_active_width - (next_key - base_key)), the key might be the same when next_key add a rand in the range.

File: db_stress_tool/db_stress_common.cc (hunk 2 above)
Comment (zhichao-cao) [reply db_stress_tool/db_stress_common.cc]: @siying .I searched the code base, some cases use std::floor() and some cases directly use floor(). It would be safer to use std::floor. I will correct that.

File: db_stress_tool/db_stress_common.cc (hunk 3 above)
Comment (siying) [reply db_stress_tool/db_stress_common.cc]: OK then.

File: db_stress_tool/db_stress_common.cc (hunk 4 above)
Comment (siying) [reply db_stress_tool/db_stress_common.cc]: Still, it used to be that every key generated is larger than the previous one.
I just wonder whether the user of the function makes assumption about the key ordering.
When you run "make crash_test", does it pass?

Comment (siying) [reply db_stress_tool/db_stress_common.cc]: Or maybe simply run db_stress with this new option.

Comment (zhichao-cao) [reply db_stress_tool/db_stress_common.cc]: @siying I have add the --hot_key_alpha=0.95 to db_crashtest.py and it passes the verification tests. I checked the code, db_crashtest.py uses an unorderset to hold the keys with no overlap. GenerateNKeys are only used in two cases: 1) Generate an array of keys for TestMultiGet, 2) Generate an array of keys as the N number of start-key for TestIterator. It is ok for them to have an array of keys that are 1) have duplication 2) the keys does not have order. To make it clear, I will add comments to GFLAGS and ahead of these function say:  the keys being generated do not ensure order (current one is always larger(smaller) than previous one).
